WordPress/gutenberg
# Ensure Query Loop Block `queryId` Uniqueness

## What?

This PR makes sure that the `queryId` for the Query Loop Block is always unique.

Closes https://github.com/WordPress/gutenberg/issues/56902

## Why?

There are multiple cases which can cause `useInstanceId()` to return IDs that are already in use:

- Duplicating a block will cause the `queryId` to be duplicated as well.
- The `instanceId` depends on the number of instances of a given block on a page. This means you can get collisions out of it if you aren't careful about how you delete Query Loop blocks and add more.

## How?
This approach relies on the fact that the `queryId` doesn't actually have to be stable, it just has to be recorded in the block's attributes so that it can be consumed. As long as it is unique and does not change during frontend page reloads it will work just fine.

`useInstanceId()` is relatively stable since it is just a counter for the number of instances of a given block on a page. This means we can safely and reliably set `queryId` on component mounting. The block won't be marked dirty because typically this value would never change unless the output of the `useInstanceId()` hook is altered due to block removal, addition, or moving.

## Testing Instructions
1. Create a post or page
2. Add multiple Query Loop blocks and note their `queryId` with the code editor.
3. Duplicate one of the blocks and confirm that the ID of the duplicated block is new.

**Note**: Switching to the code editor and back will trigger new blocks to be inserted and increment the counter. If you refresh the page with the block editor open they will re-index from 0.

### Testing Instructions for Keyboard

I'm fine writing this but looking up all of the hotkeys to switch between the code editor and duplicate blocks sounds like a lot :smile:
